hdf5-types/src/complex.rs

```diff
@@ -0,0 +1,30 @@
+use super::{CompoundField, CompoundType, H5Type, TypeDescriptor};
+
+use std::mem::size_of;
+
+use num_complex::Complex;
+
+unsafe impl<T: H5Type> H5Type for Complex<T> {
+    fn type_descriptor() -> TypeDescriptor {
+        TypeDescriptor::Compound(CompoundType {
```

Comment:
This is inherently unsafe, how do we know that offsets are right? (e.g. should we check actual offsets here?)

Reply:
I am not sure how we would check the offsets

Reply:
That would require us to make something to point at. We could make an instance of `MaybeUninit`, but field access is only a future feature

Reply:
Hm... I guess we could do something like this, does that make sense? https://rust.godbolt.org/z/TWbavj98G

Reply:
We could do
```rust
unsafe impl<T: H5Type> H5Type for Complex<T> {
    fn type_descriptor() -> TypeDescriptor {
        use std::mem::MaybeUninit;
        // MaybeUninit has the same size and is transparent, but we can
        // do some pointer arithmetics on a real instance
        let dummystruct: Complex<MaybeUninit<T>> =
            Complex::new(MaybeUninit::uninit(), MaybeUninit::uninit());

        assert_eq!(size_of_val(&dummystruct), size_of::<Complex<T>>());

        let base = addr_of!(dummystruct).cast::<u8>();
        let ptr_r = addr_of!(dummystruct.re).cast::<u8>();
        let ptr_i = addr_of!(dummystruct.im).cast::<u8>();

        TypeDescriptor::Compound(CompoundType {
            fields: vec![
                // Compatible with h5py definition of complex
                CompoundField::typed::<T>("r", unsafe { ptr_r.offset_from(base) } as usize, 0),
                CompoundField::typed::<T>("i", unsafe { ptr_i.offset_from(base) } as usize, 1),
            ],
            size: size_of::<Complex<T>>(),
        })
    }
}
```

Reply:
Might be enough to link to the documentation on representation? https://docs.rs/num-complex/0.4.2/num_complex/struct.Complex.html#representation-and-foreign-function-interface-compatibility

Reply:
Maybe add this link in a comment in code? E.g. "`Complex<T>` should be FFI-equivalent to `[T; 2]`: <link>".

Reply:
Thanks, added this as a comment under `type_descriptor()`